### ISSUE TITLE
How do you use this with `Acc` arrays?

### ISSUE THREAD

**Issue by justinlovinger** [tmcdonell/accelerate-random#4]
I have a part of a function that looks like
```haskell
sample :: Acc (Vector Double) -> IO (Acc (Vector Bool))
sample ps = do
  rs <- randomArray (uniformR (0, 1)) (A.shape ps)
  pure $ A.zipWith (A.<=) (A.use rs) ps
```

However, this doesn't work because `randomArray` expects `Shape`, but `A.shape` returns `Exp Shape`.  In other words, it looks like `mwc-random-accelerate` can't operate on `Acc` or `Exp` types.  Is there a way around this?

**Comment by tmcdonell** [comment]
Right, this package is just a shim to use mwc-random on the host, and convert that into a format Accelerate understands. But as you've found it doesn't generate the numbers in Accelerate itself.

I started a different package to try and do random number generation on the device. It's not complete yet but it might do what you want: https://github.com/tmcdonell/sfc-random-accelerate. In particular I know that if you want to get multiple random numbers in each thread it's a pain (especially if each thread needs a different number of randoms, e.g. in a loop). I have some ideas to fix it but haven't had a chance to look at it in a while. I'd be interested in knowing your use case!

P.S. sorry for the late reply!

**Comment by justinlovinger** [comment]
@tmcdonell I actually found `sfc-random-accelerate` and have been using it instead of `mwc-random-accelerate`.

I'm writing a mathematical optimization library with `accelerate` for arrays.  Most derivative-free optimizers, and some derivative optimizers, make heavy use of random numbers; genetic algorithms and PBIL for example.

`sfc-random-accelerate` is working well.  My main complaint is that the shape of a random array is tied to the generator.  I had to do some refactoring to make sure a user can't accidentally pass a generator that doesn't match the shape needed by an optimizer with a particular state.  Also, managing multiple generators isn't ideal, if I need random arrays of different shapes within the same algorithm.
